uqcsbot/scripts/wolfram.py

```diff
@@ -0,0 +1,192 @@
+
+
+@bot.on_command("wolfram")
+async def handle_wolfram(command: Command):
+    """
+    This uses wolfram's conversation api to return a short response that can be replied to in a thread.
+    If the response cannot be replied to a general short answer response is displayed instead.
+
+    Example Usage:
+    !wolfram Solve Newton's Second Law for mass
+    !wolfram What is the distance from Earth to Mars?
+
+    and then start a thread to continue to conversation
+    """
+    api_url = r"http://api.wolframalpha.com/v1/conversation.jsp"
+    search_query = command.arg
+    http_response = await bot.run_async(requests.get, api_url, params={'input': search_query, 'appid': APP_ID})
+
+    # Check if the response is okay
+    if http_response.status_code != requests.codes.ok:
+        bot.post_message(command.channel, "There was a problem getting the response")
+        return
+
+    wolfram_result = json.loads(http_response.content)
+    if 'error' in wolfram_result:
+        error = wolfram_result['error']
+        if error == "No result is available":
+            # If no conversational result is available just return a normal short answer
+            short_response = await short_answer(search_query)
+            bot.post_message(command.channel, short_response)
+            return
+        else:
+            bot.post_message(command.channel, error)
+            return
+
+    # Unpack the response in order to construct the reply message
+    result = wolfram_result['result']  # This is the answer to our question
+    reply_host = wolfram_result['host']  # This is the hostname to ask the next question to
+    conversation_id = wolfram_result['conversationID']  # Used to continue the conversation
+    s_output = wolfram_result.get('s', None)  # s is only sometimes returned but is vital if it is returned
+
+    # TODO: Is there a better option than storing the id in the fallback?
+    # Here we store the conversation ID in the fallback so we can get it back later.
+    # We also store an identifier string to check against later and the reply_host and s_output string
+    attachments = [{
+        'fallback': f'WolframCanReply {reply_host} {s_output} {conversation_id}',
+        'footer': 'Further questions may be asked',
+        'text': result,
+    }]
+
+    bot.post_message(command.channel, "", attachments=attachments)
+
+
+@bot.on('message')
+async def handle_reply(evt: dict):
+    # If the message isn't from a thread or is from a bot ignore it (avoid those infinite loops)
+    if 'thread_ts' not in evt or (evt.get('subtype') == 'bot_message'):
+        return
+
+    channel = evt['channel']
+    thread_ts = evt['thread_ts']  # This refers to time the original message
+    thread_parent = bot.api.conversations.history(channel=channel, limit=1, inclusive=True, latest=thread_ts)
+
+    if not thread_parent['ok']:
+        bot.post_message(channel, 'Sorry, something went wrong with slack', thread_ts=thread_ts)
```

Comment:
so what does this section do? does it post to the thread? doesn't being able to post to the thread assume that there's a thread parent, which we just learnt wasn't ok?

Reply:
This is just a check to ensure that the thread parent was returned. I was running into a lot of auth issues in my test slack so if we see this come up that is the likely culprit.